crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,301 @@
+type FetchRequest = (OpId, AgentId);
+
+#[derive(Debug)]
+struct State {
+    ops: HashSet<FetchRequest>,
+    cool_down_list: HashMap<AgentId, Instant>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    state: Arc<Mutex<State>>,
+    fetch_queue_tx: Sender<FetchRequest>,
+    fetch_queue_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+}
+
+impl Inner {
+    pub fn new() -> Self {
+        // Create a channel to send new ops to fetch to the tasks. This is in effect the fetch queue.
+        let (fetch_queue_tx, fetch_queue_rx) = channel::<FetchRequest>(1024);
+        let fetch_queue_rx = Arc::new(Mutex::new(fetch_queue_rx));
+
+        let ops = HashSet::new();
+        let cool_down_list = HashMap::new();
+
+        Self {
+            state: Arc::new(Mutex::new(State {
+                ops,
+                cool_down_list,
+            })),
+            fetch_queue_tx,
+            fetch_queue_rx,
+        }
+    }
+
+    pub fn create_fetch_queue(
+        &self,
+        config: CoreFetchConfig,
+        transport: DynTransport,
+    ) {
+        for _ in 0..config.parallel_request_count {
+            tokio::task::spawn(Inner::fetch_task(
+                self.fetch_queue_tx.clone(),
+                self.fetch_queue_rx.clone(),
+                transport.clone(),
+                self.state.clone(),
+                config.cool_down_interval_ms,
+            ));
+        }
+    }
+
+    async fn fetch_task(
+        fetch_request_tx: Sender<FetchRequest>,
+        fetch_request_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+        transport: DynTransport,
+        state: Arc<Mutex<State>>,
+        cool_down_interval: u64,
+    ) {
+        while let Some((op_id, agent_id)) =
+            fetch_request_rx.lock().await.recv().await
+        {
+            // Ensure op is still in the set of ops to fetch.
+            if !state
+                .lock()
+                .await
+                .ops
+                .contains(&(op_id.clone(), agent_id.clone()))
+            {
+                continue;
+            }
+
+            // Check if agent is on cool-down list.
+            if !Inner::is_agent_cooling_down(
+                &agent_id,
+                &mut state.lock().await.cool_down_list,
+                cool_down_interval,
+            ) {
+                // Send fetch request to agent.
+                if let Err(err) = transport
+                    .lock()
+                    .await
+                    .send_op_request(op_id.clone(), agent_id.clone())
+                    .await
+                {
+                    eprintln!("could not send fetch request for op {op_id} to agent {agent_id}: {err}");
+                    state
+                        .lock()
+                        .await
+                        .cool_down_list
+                        .insert(agent_id.clone(), Instant::now());
+                }
+            }
+
+            // Re-insert the fetch request into the queue.
+            if let Err(err) = fetch_request_tx
```

Comment:
As above, this reads to me like it can deadlock. If gossip discovers a batch of ops and fills the channel, then both fetch tasks experience an error, this won't be able to send and it is also responsible for draining the channel.

Reply:
done https://github.com/holochain/kitsune2/commit/d13f217f86acb8715a3e1fb314e33d83012b9091